var/spack/repos/builtin/packages/cvise/package.py

```diff
@@ -33,3 +35,7 @@ class Cvise(CMakePackage):
 
     depends_on("py-pytest", when="+pytest", type=("build", "run"))
     depends_on("colordiff", when="+colordiff", type=("build", "run"))
+
+    # LLVM depends on but doesn't provide terminfo CMake targets and C-Vise
+    # fails during configuration
+    conflicts("ncurses +termlib")
```

Comment:
This has to be `conflicts("^ncurses +termlib")` (note the prefix of `^`).

Reply:
I actually started off with `^ncurses` and went to `ncurses` once I tested the concretization of `cvise` on another system. I'm definitely not saying `ncurses` is right because it's just my workaround to the consequences I saw :upside_down_face: When I had `^ncurses` I get a concretization like this:
```
> spack spec -lI cvise
Input spec
--------------------------------
 -   cvise

Concretized
--------------------------------
 -   kvl5oca  cvise@2.10.0%gcc@13.2.0~colordiff~ipo~pytest build_system=cmake build_type=Release generator=make arch=linux-nixos24-icelake
[e]  sdscwht      ^cmake@3.28.3%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-nixos24-icelake
 -   glgzdni      ^flex@2.6.3%gcc@13.2.0+lex~nls build_system=autotools arch=linux-nixos24-icelake
 -   fyhjkmm          ^bison@3.8.2%gcc@13.2.0~color build_system=autotools arch=linux-nixos24-icelake
[e]  j3v67vx          ^diffutils@3.10%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
[e]  u2emupb          ^findutils@4.9.0%gcc@13.2.0 build_system=autotools patches=440b954 arch=linux-nixos24-icelake
[+]  xj5lb5m          ^m4@1.4.19%gcc@13.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-nixos24-icelake
[+]  gnrdfxf              ^libsigsegv@2.14%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
[+]  ovjdi6x      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-nixos24-icelake
[e]  yu4id25      ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-nixos24-icelake
 -   ltmafwu      ^llvm@18.1.1%gcc@13.2.0+clang~cuda~flang+gold~ipo+libomptarget~libomptarget_debug~link_llvm_dylib+lld+lldb+llvm_dylib+lua~mlir+polly~python~split_dwarf~z3~zstd build_system=cmake build_type=Release compiler-rt=runtime generator=ninja libcxx=runtime libunwind=runtime openmp=runtime shlib_symbol_version=none targets=all version_suffix=none arch=linux-nixos24-icelake
[e]  y2we6f3          ^binutils@2.41%gcc@13.2.0~gas+gold~gprofng+headers~interwork+ld~libiberty~lto~nls~pgo+plugins build_system=autotools compress_debug_sections=zlib libs=shared,static arch=linux-nixos24-icelake
[+]  hc74rpm          ^hwloc@2.9.1%gcc@13.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-nixos24-icelake
[+]  c3n5rrg              ^libpciaccess@0.17%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
[+]  a2kx342                  ^libtool@2.4.7%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
[+]  xkhsgnx                  ^util-macros@1.19.3%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
 -   txddtpd          ^libedit@3.1-20230828%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
 -   7bnhvdu          ^libffi@3.4.4%gcc@13.2.0 build_system=autotools patches=070b1f3 arch=linux-nixos24-icelake
[+]  kbrbdow          ^libxml2@2.10.3%gcc@13.2.0+pic~python+shared build_system=autotools arch=linux-nixos24-icelake
[+]  tz2ncke              ^libiconv@1.17%gcc@13.2.0 build_system=autotools libs=shared,static arch=linux-nixos24-icelake
 -   ao5o6kq          ^lua@5.3.6%gcc@13.2.0~pcfile+shared build_system=makefile fetcher=curl arch=linux-nixos24-icelake
[e]  r6nx5yi              ^curl@8.6.0%gcc@13.2.0+gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-nixos24-icelake
 -   7fnodda              ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-nixos24-icelake
 -   nxmegk3              ^unzip@6.0%gcc@13.2.0 build_system=makefile patches=f6f6236 arch=linux-nixos24-icelake
[+]  zoexrvh          ^ncurses@6.4%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-nixos24-icelake
[e]  xvxfyxr          ^ninja@1.11.1%gcc@13.2.0+re2c build_system=generic arch=linux-nixos24-icelake
 -   cwpwly2          ^perl-data-dumper@2.173%gcc@13.2.0 build_system=perl arch=linux-nixos24-icelake
[e]  7cgf6x7              ^perl@5.38.2%gcc@13.2.0~cpanm+opcode+open+shared+threads build_system=generic arch=linux-nixos24-icelake
[+]  fvsrq4p          ^pkgconf@1.9.5%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
 -   l6oh25a          ^swig@4.1.1%gcc@13.2.0 build_system=autotools arch=linux-nixos24-icelake
 -   ocqghts              ^pcre2@10.42%gcc@13.2.0~jit+multibyte build_system=autotools arch=linux-nixos24-icelake
[+]  7j7gpbo          ^xz@5.4.6%gcc@13.2.0~pic build_system=autotools libs=shared,static arch=linux-nixos24-icelake
[+]  ut34que          ^zlib-ng@2.1.5%gcc@13.2.0+compat+opt build_system=autotools arch=linux-nixos24-icelake
 -   sfhnikq      ^py-chardet@5.2.0%gcc@13.2.0 build_system=python_pip arch=linux-nixos24-icelake
 -   n4arrsg          ^py-pip@23.1.2%gcc@13.2.0 build_system=generic arch=linux-nixos24-icelake
 -   cxx2uvc          ^py-setuptools@69.2.0%gcc@13.2.0 build_system=generic arch=linux-nixos24-icelake
 -   jhx3qfu          ^py-wheel@0.41.2%gcc@13.2.0 build_system=generic arch=linux-nixos24-icelake
 -   dwalhsh      ^py-pebble@5.0.3%gcc@13.2.0 build_system=python_pip arch=linux-nixos24-icelake
 -   ov73o4n      ^py-psutil@5.9.5%gcc@13.2.0 build_system=python_pip arch=linux-nixos24-icelake
[e]  agtj2tj      ^python@3.11.8%gcc@13.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-nixos24-icelake
 -   3qy7vg4      ^unifdef@2.12%gcc@13.2.0 build_system=makefile arch=linux-nixos24-icelake
```

Note the `[+]  zoexrvh          ^ncurses@6.4%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-nixos24-icelake`, which is definitely not what I want. I changed it to `conflicts("ncurses +termlib", ...)` and got the concretization I expected but that may of course just be a concidence. `ncurses` is not a direct dependency of `cvise` and in principle doesn't care about `ncurses`. I would also not like to add a conflict to llvm, because llvm itself builds just fine even with `ncurses +termlib`, the issue only shows up in this case when configuring `cvise`.

Do you think this could be a bug in the concretizer? Or is there another way of expressing the conflict so that I actually get an `ncurses ~termlib` in this concretization?

Reply:
The following should satisfy both this requirement and the audit:

```
depends_on("ncurses ~termlib", when="^llvm+lldb")
```

a bit awkward, but if the `cvise` install is asking questions about ncurses, then I think it effectively depends on it.

@alalazo this is a bit awkward but we don't have a way to constrain "exclusively-transitive" dependencies: those that show up only via another dependency, I'm curious if you had any thoughts on this (mainly since you appear to have written some of the code involved in this audit failure). For example, a slightly-more-robust version of the above would be something like:

```
conflicts("^ncurses +termlib")
```

or at least that gets close to what I *want* to say - which is: "disable ~termlib on ncurses if it appears anywhere as a transitive dependency") - but this doesn't actually do anything (since `cvise` doesn't actually directly depend on `ncurses`).

Reply:
Thanks @scheibelp! That gives me the concretization I hoped for and I've pushed an update (with a slightly more verbose comment on the `depends_on`). I left out the `+lldb` from `when=^llvm` since I assumed that was spurious. Let me know if you'd like that added back.

Reply:
>  I left out the +lldb from when=^llvm since I assumed that was spurious.

I meant to add it: llvm only depends on `ncurses` when the `+lldb` variant is enabled on it. Without this, `cvise` will depend on `ncurses` even if `llvm` does not (which I didn't think was required).

(apologies for late response)

Reply:
No problem! That makes sense, I've added `+lldb` back.